providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/core_numbers.h>
+#include <openssl/core_names.h>
+#include <openssl/bn.h>
+#include <openssl/ec.h>
+#include <openssl/objects.h>
+#include <openssl/params.h>
+#include "internal/param_build.h"
+#include "prov/implementations.h"
+#include "prov/providercommon.h"
+
+static OSSL_OP_keymgmt_importdomparams_fn ec_importdomparams;
+static OSSL_OP_keymgmt_exportdomparams_fn ec_exportdomparams;
+static OSSL_OP_keymgmt_importkey_fn ec_importkey;
+static OSSL_OP_keymgmt_exportkey_fn ec_exportkey;
+static OSSL_OP_keymgmt_query_operation_name_fn ec_query_operation_name;
+
+static
+const char *ec_query_operation_name(int operation_id)
+{
+    switch (operation_id) {
+    case OSSL_OP_KEYEXCH:
+        return "ECDH";
+#if 0
+    case OSSL_OP_SIGNATURE:
```

Comment:
Talking about the support of SM2 is based on the EC part has been done, is it suggested that SM2 should be implemented with ec_kmgmt or should SM2 have an sm2_kmgmt?

Reply:
It should have it's dedicated keymgmt so that it can have a specific key generation algorithm!

Reply:
But to avoid code duplication it might maybe be defined in the same .c file and just defines the alternative keygen method.

Reply:
Considering the relationship between SM2 key and a normal EC key, I think it's impossible to separate the implementations of them. For instance, if one day SM2 signature algorithm is moved to a national provider but without any EC code in it, then that provider needs to depend on the default provider to use the EC code inside it.

But if a national provider provides an SM2 implementation without depending on the OpenSSL EC code, that would be totally another story...

Reply:
What I mean is that it should define its own keymgmt object inside this file and resume most of the functions currently defined.

It's going to be mostly a clone of the EC keymgmt object but it will define a different function for the key generation and declare a different signature object.

This is because when generating a key they use a different algorithm than normal EC keys as the range of valid secret scalars do not overlap.

But they will share the same encoding, decoding, export and import code.

Reply:
EC and SM2 will have to have different KEYMGMT implementations insofar that the OSSL_ALGORITHM will need to use distinct names, and that the OSSL_DISPATCH they point at will possibly have to not be exactly the same.  But other than that, *nothing* stops the OSSL_DISPATCH arrays for each of them to largely use the exact same functions (I imagine that the init function may have to be different, to set some kind of flag indicating which kind of key we're dealing with)

Reply:
I just updated the description to include notes I took so far about EC and SM2.

Reply:
> What I mean is that it should define its own keymgmt object inside this file and reuse most of the functions currently defined.

Okay. If doing so, the problem is the OpenSSL SM2 will probably not be possible to be in another provider other than the default provider. Is the whole provider mechanism designed to allow one provider to depend on the default provider?

Reply:
Well, not sure in general about dependency among providers, but almost surely it would not be possible to go beyond the level of granularity of a `OSSL_ALGORITHM`.

So no, if they are in 2 different providers code duplication is almost inevitable.

Unless they are both part of this repo and we can do `ifdef` tricks:
 e.g. we include the ec_keymgmt.c file in the default provider, in the FIPS provider and in the Chinese provider included in the openssl code base; when compiling the first 2 a preprocessor `#define` selects the definition of the EC `OSSL_DISPATCH`, when compiling the Chinese provider the `ifdef` selects the definitions relevant to define the SM2 `OSSL_DISPATCH`.